File: apps/api/app/schemas/election.py
# ############################################################################
# AI_HEADER: MODULE_SCHEMAS_ELECTION
# ROLE: Pydantic schemas for election endpoints
# DEPENDENCIES: pydantic, app.schemas.camel
# GRACE_ANCHORS: []
# ############################################################################

# START_MODULE_CONTRACT: M-SCHEMAS-ELECTION
# purpose: Request and response models for /api/election routes.
# owns:
#   - apps/api/app/schemas/election.py
# inputs: request dicts
# outputs: validated models
# dependencies: none
# side_effects: none
# emitted_logs: none
# invariants: none
# failure_policy: none
# END_MODULE_CONTRACT: M-SCHEMAS-ELECTION

# START_MODULE_MAP: M-SCHEMAS-ELECTION
# public_entrypoints:
#   - ElectionDayNote
#   - ElectionSearchCreateRequest
#   - ElectionSearchRead
# semantic_blocks: none
# owned_tests: none
# END_MODULE_MAP: M-SCHEMAS-ELECTION

from __future__ import annotations

from datetime import date, datetime
import uuid
from typing import Any, Literal
from pydantic import BaseModel, Field

from app.schemas._base import CamelModel


class ElectionDayNote(BaseModel):
    date: str
    note: str


class ElectionNarrative(BaseModel):
    hero_reason: str
    hero_personal: str
    hero_plain: str
    hero_hours: str
    day_notes: list[ElectionDayNote]
    avoid_notes: list[ElectionDayNote]


def validate_election_narrative(
    narrative_data: dict[str, Any],
    expected_best_dates: list[str],
    expected_avoid_dates: list[str],
    fallback_notes: dict[str, str] | None = None,
) -> ElectionNarrative:
    # START_FUNCTION_CONTRACT: F-M-SCHEMAS-ELECTION.validate_election_narrative
    # purpose: Validate LLM output against ElectionNarrative schema and dates alignment.
    # inputs: narrative_data (dict), expected_best_dates (list[str]),
    #   expected_avoid_dates (list[str]), fallback_notes (dict date->note from
    #   engine facts, used to fill dates a weak model omitted — never fabricated).
    # returns: ElectionNarrative with notes aligned to expected dates in expected order.
    # error_behavior: raises ValueError on schema error or unknown dates in notes.
    # END_FUNCTION_CONTRACT: F-M-SCHEMAS-ELECTION.validate_election_narrative
    try:
        parsed = ElectionNarrative.model_validate(narrative_data)
    except Exception as exc:
        raise ValueError(f"Election narrative schema validation failed: {exc}") from exc

    def _align(notes: list[ElectionDayNote], expected: list[str], kind: str) -> list[ElectionDayNote]:
        unknown = [n.date for n in notes if n.date not in expected]
        if unknown:
            raise ValueError(f"Election {kind} contain unknown dates {unknown}, expected {expected}")
        by_date = {n.date: n for n in notes}
        aligned: list[ElectionDayNote] = []
        for d in expected:
            if d in by_date:
                aligned.append(by_date[d])
            elif fallback_notes and d in fallback_notes:
                aligned.append(ElectionDayNote(date=d, note=fallback_notes[d]))
            else:
                raise ValueError(f"Election {kind} missing required date {d}")
        return aligned

    return ElectionNarrative(
        hero_reason=parsed.hero_reason,
        hero_personal=parsed.hero_personal,
        hero_plain=parsed.hero_plain,
        hero_hours=parsed.hero_hours,
        day_notes=_align(parsed.day_notes, expected_best_dates, "day_notes"),
        avoid_notes=_align(parsed.avoid_notes, expected_avoid_dates, "avoid_notes"),
    )


class ElectionSearchCreateRequest(CamelModel):
    event_type: str = Field(..., description="Event type identifier")
    window_from: date = Field(..., description="Start date (inclusive)")
    window_to: date = Field(..., description="End date (inclusive)")
    idempotency_key: str = Field(..., max_length=80, description="Idempotency key")
    client_timezone: str | None = Field(None, description="Client timezone string")


class ElectionDayReasonRead(CamelModel):
    date: date
    score: int
    label: Literal["great", "good", "ok", "avoid"]
    reasons: list[str]


class ElectionResultRead(CamelModel):
    event: str
    best_days: list[ElectionDayReasonRead]
    avoid_days: list[ElectionDayReasonRead]


class ElectionSearchRead(CamelModel):
    id: uuid.UUID
    event_type: str
    window_from: date
    window_to: date
    status: Literal["pending", "processing", "done", "failed", "refunded"]
    created_at: datetime
    result: dict[str, Any] | None = None
    public_error_code: str | None = None
    public_error_message: str | None = None

File: lib/contracts/election.ts
// ############################################################################
// AI_HEADER: MODULE_CONTRACTS_ELECTION
// ROLE: UI — election contracts & zod schemas v2
// DEPENDENCIES: zod
// ############################################################################

import { z } from "zod"

export type ElectionSubCategory = {
  key: string
  label: string
}

export type ElectionCategory = {
  key: string
  label: string
  emoji: string
  subs: ElectionSubCategory[]
}

export const ELECTION_CATEGORIES: readonly ElectionCategory[] = [
  {
    key: "relations",
    label: "Отношения",
    emoji: "💕",
    subs: [
      { key: "date", label: "Свидание" },
      { key: "wedding", label: "Свадьба/помолвка" },
      { key: "reconcile", label: "Примирение" },
    ],
  },
  {
    key: "work",
    label: "Работа и дела",
    emoji: "💼",
    subs: [
      { key: "job", label: "Новая работа" },
      { key: "launch", label: "Запуск проекта" },
      { key: "contract", label: "Документы и договоры" },
    ],
  },
  {
    key: "travel",
    label: "Поездки",
    emoji: "🛫",
    subs: [
      { key: "leisure", label: "Путешествие/отдых" },
      { key: "business", label: "Деловая поездка" },
      { key: "family", label: "Поездка к родным" },
    ],
  },
  {
    key: "money",
    label: "Покупки и финансы",
    emoji: "🛍",
    subs: [
      { key: "purchase", label: "Крупная покупка" },
      { key: "invest", label: "Вложения и финрешения" },
    ],
  },
  {
    key: "beauty",
    label: "Красота и тело",
    emoji: "💇‍♀️",
    subs: [
      { key: "salon", label: "Стрижка/салон" },
      { key: "treatment", label: "Процедуры и лечение" },
    ],
  },
  {
    key: "family",
    label: "Семья и дом",
    emoji: "👶",
    subs: [
      { key: "conception", label: "Зачатие ребёнка" },
      { key: "moving", label: "Переезд/новоселье" },
    ],
  },
] as const

export const ElectionDayFactSchema = z.object({
  date: z.string(),
  score: z.number(),
  label: z.enum(["great", "good", "ok", "avoid"]),
  reasons: z.array(z.string()),
  moon_sign: z.string().optional(),
  moon_sign_ru: z.string().optional(),
  waxing: z.boolean().optional(),
  phase_pct: z.number().optional(),
  voc_fraction: z.number().optional(),
  voc_intervals: z.array(z.string()).optional(),
  mercury_retro: z.boolean().optional(),
})

export type ElectionDayFact = z.infer<typeof ElectionDayFactSchema>

export const ElectionNarrativeNoteSchema = z.object({
  date: z.string(),
  note: z.string(),
})

export const ElectionNarrativeSchema = z.object({
  hero_reason: z.string(),
  hero_personal: z.string(),
  hero_plain: z.string(),
  hero_hours: z.string(),
  day_notes: z.array(ElectionNarrativeNoteSchema),
  avoid_notes: z.array(ElectionNarrativeNoteSchema),
})

export type ElectionNarrative = z.infer<typeof ElectionNarrativeSchema>

export const ElectionResultFactsSchema = z.object({
  event: z.object({
    category: z.string().optional(),
    sub: z.string().optional(),
    label: z.string().optional(),
  }).optional(),
  personal: z.object({
    natal_moon_sign: z.string().nullable().optional(),
    natal_moon_sign_ru: z.string().nullable().optional(),
    resonates: z.boolean().optional(),
  }).optional(),
})

export type ElectionResultFacts = z.infer<typeof ElectionResultFactsSchema>

export const ElectionResultSchema = z.object({
  event: z.string(),
  best_days: z.array(ElectionDayFactSchema),
  avoid_days: z.array(ElectionDayFactSchema),
  days: z.array(ElectionDayFactSchema).optional(),
  facts: ElectionResultFactsSchema.optional(),
  narrative: ElectionNarrativeSchema.optional(),
})

export type ElectionResult = z.infer<typeof ElectionResultSchema>

export const ElectionSearchSchema = z.object({
  id: z.string().uuid(),
  eventType: z.string(),
  windowFrom: z.string(),
  windowTo: z.string(),
  status: z.enum(["pending", "processing", "done", "failed", "refunded"]),
  createdAt: z.string(),
  result: ElectionResultSchema.nullable().optional(),
  publicErrorCode: z.string().nullable().optional(),
  publicErrorMessage: z.string().nullable().optional(),
})

export type ElectionSearch = z.infer<typeof ElectionSearchSchema>
